irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -15,103 +15,118 @@
+            [this](auto block) {
+              static const auto in_proposal = [this](const auto &tx) {
+                return this->proposal_set_.count(hash(tx).to_string());
+              };
+              boost::for_each(
+                  block.transactions | boost::adaptors::filtered(in_proposal),
+                  notify_success);
             },
             // on complete
             [this]() {
-              for (auto tx_hash : proposal_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status =
-                    TransactionResponse::STATEFUL_VALIDATION_FAILED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              proposal_set_.clear();
-
-              for (auto tx_hash : candidate_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status = TransactionResponse::COMMITTED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              candidate_set_.clear();
+              boost::for_each(
+                  boost::join(this->proposal_set_, this->candidate_set_),
+                  notify_fail);
+              this->proposal_set_.clear();
+              this->candidate_set_.clear();
             });
       });
+
+      mst_proc_->onPreparedTransactions().subscribe(
+          [](auto tx) { return notify_success(*tx); });
+      mst_proc_->onExpiredTransactions().subscribe(
+          [](auto tx) { return notify_fail(hash(*tx).to_string()); });
     }
 
     void TransactionProcessorImpl::transactionHandle(
-        std::shared_ptr<model::Transaction> transaction) {
+        ConstRefTransaction transaction) {
       log_->info("handle transaction");
-      model::TransactionResponse response;
-      response.tx_hash = hash(*transaction).to_string();
-      response.current_status =
-          model::TransactionResponse::Status::STATELESS_VALIDATION_FAILED;
+      TransactionResponse response{hash(*transaction).to_string(),
+                                   Status::STATELESS_VALIDATION_FAILED};
 
+      // TODO: nice place for code linearizing
       if (validator_->validate(*transaction)) {
-        response.current_status =
-            TransactionResponse::Status::STATELESS_VALIDATION_SUCCESS;
-        pcs_->propagate_transaction(transaction);
+        response.current_status = Status::STATELESS_VALIDATION_SUCCESS;
+        switch (transaction->signatures.size()) {
+          case 0:
+            response.current_status = Status::STATELESS_VALIDATION_FAILED;
+            break;
+          case 1:
+            pcs_->propagate_transaction(transaction);
+            break;
+          default:
+            mst_proc_->propagateTransaction(transaction);
```

Comment:
`propagateTransaction` not so suitable name. Maybe refactor it?

Reply:
Imo we should refactor all the torii-related stuff in separate pr, it will be quite big

Reply:
I mean `propagateTransaction` from MST.